### Задача 1

Дана матрица объект-признак
```
X = [[ 1.22401313,  2.30868478,  3.03636353,  2.69287214],
     [-0.18757272,  1.30337355,  5.12093014,  3.46363202],
     [-0.81094525,  1.82463398,  5.79686488,  1.86159445],
     [ 0.75129018,  2.67392052,  3.65529809,  1.66746094],
     [ 0.00972362,  1.97367255,  2.50594319,  1.69755173],
     [-0.62972637,  0.77750764,  2.84124027,  4.54410559],
     [ 2.29536229,  1.81206697,  1.95026215,  1.51874636],
     [ 0.0920418 ,  2.26971361,  7.47708735,  2.61081203],
     [ 2.39252799,  3.17563985,  3.61420599,  5.10773362],
     [ 0.54983815,  2.87988651,  1.65752765,  1.59635987]]
```
и значения целевой переменной
```
y = [ 9.26193358,  9.700363  ,  8.67214805,  8.74796974,  6.18689108,
      7.53312713,  7.57643777, 12.44965478, 14.29010746,  6.68361218]
```

Подберите два признака (из четырёх) так, чтобы уровень линейной зависимости целевой переменной от значений этих признаков был максимальным. Другими словами, модель линейной регрессии на этих признаках должна давать наилучший результат. В качестве ответа запишите значение коэффициента детерминации для модели на этих признаках.


In [1]:
import numpy as np

In [2]:
X = np.array([[ 1.22401313,  2.30868478,  3.03636353,  2.69287214],
     [-0.18757272,  1.30337355,  5.12093014,  3.46363202],
     [-0.81094525,  1.82463398,  5.79686488,  1.86159445],
     [ 0.75129018,  2.67392052,  3.65529809,  1.66746094],
     [ 0.00972362,  1.97367255,  2.50594319,  1.69755173],
     [-0.62972637,  0.77750764,  2.84124027,  4.54410559],
     [ 2.29536229,  1.81206697,  1.95026215,  1.51874636],
     [ 0.0920418 ,  2.26971361,  7.47708735,  2.61081203],
     [ 2.39252799,  3.17563985,  3.61420599,  5.10773362],
     [ 0.54983815,  2.87988651,  1.65752765,  1.59635987]])
y = np.array([ 9.26193358,  9.700363  ,  8.67214805,  8.74796974,  6.18689108,
      7.53312713,  7.57643777, 12.44965478, 14.29010746,  6.68361218])

#### Для начала повторим процедуру расчетов

In [3]:
ones = np.ones((X.shape[0], 1)) #создаем еденичный вектор по количесту наблюдений
X = np.hstack((ones, X)) #Добавляем вектор в признаковое пространство

In [4]:
XTX = X.T.dot(X) #X.транспонированное.матрично умножаем на X изначальное

In [5]:
XTX_inv = np.linalg.inv(XTX) # Ищем обратную матрицу предыдущего расчета

In [6]:
b = XTX_inv.dot(X.T).dot(y) # Эту обратную матрицу.матрично умножаем на транспонированную матрицу X.матрично умножаем на Y

In [7]:
b # Полученные коэфф. [b0 , b1, b2, b3, b4]

array([-7.50726770e-09,  1.00000000e+00,  1.00000000e+00,  9.99999999e-01,
        1.00000000e+00])

In [8]:
z = X.dot(b) # спрогнозируем

In [9]:
e = (y-z) # получим остатки (ошибку) на каждом прогнозе

In [10]:
e.mean() # средняя ошибка стремится к 0

-2.2115642650533117e-14

In [11]:
def sum_of_squares(samples: np.ndarray) -> float:
    """Сумма квадратов отклонений.
    """
    
    return ((samples - samples.mean()) ** 2).sum()

In [12]:
R = 1 - sum_of_squares(e) / sum_of_squares(y)

In [13]:
R

1.0

Ага, R = 1, мечта Датасаинтиста..

##### Напишем функционал для нашей задачи.

In [14]:
x = np.array([[ 1.22401313,  2.30868478,  3.03636353,  2.69287214],
     [-0.18757272,  1.30337355,  5.12093014,  3.46363202],
     [-0.81094525,  1.82463398,  5.79686488,  1.86159445],
     [ 0.75129018,  2.67392052,  3.65529809,  1.66746094],
     [ 0.00972362,  1.97367255,  2.50594319,  1.69755173],
     [-0.62972637,  0.77750764,  2.84124027,  4.54410559],
     [ 2.29536229,  1.81206697,  1.95026215,  1.51874636],
     [ 0.0920418 ,  2.26971361,  7.47708735,  2.61081203],
     [ 2.39252799,  3.17563985,  3.61420599,  5.10773362],
     [ 0.54983815,  2.87988651,  1.65752765,  1.59635987]])
y = np.array([ 9.26193358,  9.700363  ,  8.67214805,  8.74796974,  6.18689108,
      7.53312713,  7.57643777, 12.44965478, 14.29010746,  6.68361218])

In [15]:
def test_feachure(x: np.ndarray, y: np.ndarray, cols: list ):
    """
    x - пространство признаков
    y - вектор ответов
    cols - список используемых фич.
    """
    x = x[:,cols]
    ones = np.ones((x.shape[0], 1)) 
    x = np.hstack((ones, x))
    XTX = x.T.dot(x)
    XTX_inv = np.linalg.inv(XTX)
    b = XTX_inv.dot(x.T).dot(y)
    z = x.dot(b)
    e = (y-z)
    R = 1 - sum_of_squares(e) / sum_of_squares(y)
    return R

In [16]:
from itertools import combinations

In [17]:
for i in list(combinations([0,1,2,3], 2)):
    print(f"{list(i)} - {test_feachure(x,y,list(i))}")

[0, 1] - 0.18113594742585215
[0, 2] - 0.7634246238793152
[0, 3] - 0.4532966783144077
[1, 2] - 0.547948273403901
[1, 3] - 0.6062055761129932
[2, 3] - 0.622441987650532


#### Ответ: Столбцы - [0, 2],  Коэфф детерминации - 0.7634246238793152

### Задача 2.
Является ли значимым уравнение регрессии, полученное в предыдущей задаче? В качестве ответа запишите 1, если является, и 0 иначе.


* Считать будем по тому уравнению, в котором 2 признака дают наилучший результат детерминации.

In [18]:
import scipy.stats as stats


- H0 - Зависимость случайна
- Н1 - Зависимость значима
- Используем распределение Фишера
- alpha = 0.05
- Критическая область правосторонняя

In [19]:
R2 = test_feachure(x,y,[0, 2])
m = 2
n = 10
alpha = 0.05
k1 = m
k2 = n - m - 1

In [20]:
F = (R2/k1) / ((1-R2)/k2)

In [21]:
t = stats.f.ppf(1-alpha, k1, k2)

In [22]:
F, t

(11.29443912292265, 4.73741412777588)

F - наша статистика больше t, значит лежит в критической области. Таким образом мы принимаем гипотезу Н1 - уравнение значимо.

#### Ответ: 1 - значимо

### Задача 3.
Для проведения A/B-тестирования сайта интернет-магазина были получены следующие данные: страница `A` была посещена 2509 раз, из них 77 закончились совершением покупки, страница `B` была посещена 1465 раз, 60 из них закончились совершением покупки. Является ли значимым отличие конверсии на страницах `A` и `B`? В качестве ответа запишите 1, если является, и 0 иначе.

_Подсказка_. Реализуйте двухвыборочный t-тест. В качестве выборок здесь можно взять наборы меток совершения покупки (0 или 1) каждым посетителем.

In [23]:
A = np.array([1]*77 + [0]*(2509-77))
B = np.array([1]*60 + [0]*(1465-60))

In [24]:
nA = A.size
nB = B.size
stdA = A.std(ddof=1)
stdB = B.std(ddof=1)

In [25]:
deltastd = (((stdA**2) / nA) + ((stdB**2) / nB))**0.5

In [26]:
t = (A.mean() - B.mean()) / deltastd

In [27]:
t

-1.6504551408398205

In [28]:
df = (stdA ** 2 / nA + stdB ** 2 / nB) ** 2 / \
    ((stdA ** 2 / nA) ** 2 / (nA - 1) + (stdB ** 2 / nB) ** 2 / (nB - 1))

df

2732.8025644352133

In [29]:
t1 = stats.t.ppf(alpha / 2, df=df)
t2 = stats.t.ppf(1 - alpha / 2, df=df)

t1, t2

(-1.9608324352746576, 1.9608324352746571)

Мы видим, что наша метрика t не выходит за критическую область, следовательно H0 - принимается. Конверсия не является значимой.

In [30]:
stats.ttest_ind(A, B, equal_var=False)

Ttest_indResult(statistic=-1.6504551408398205, pvalue=0.09896479217925898)

#### Ответ: 0